File: src/main/scala/perft/PerftSpeedTest.scala
package perft

import java.io.IOException

import position._

import scala.collection.Iterator
import scala.collection.mutable.ListBuffer

object PerftSpeedTest {
  @throws[IOException]
  def main(args: Array[String]) {
    perftTest
  }

  def perftTest() {
    val f: String = ICodage.FEN_INITIALE
    val gp: GPositionS = FenToGPosition.toGPosition(f)
    val max_depth: Int = 6
    val t0: Double = System.nanoTime
    var depth: Int = 1
    while (depth <= max_depth) {
      {
        val res = perft(gp, depth)
        val t1: Double = System.nanoTime
        System.out.println("Depth " + depth + " : " + (t1 - t0) / 1000000000 + " sec")
        System.out.println("Count = " + res.moveCount)
      }
      {
        depth += 1
      }
    }
  }

  def perft(gp: GPositionS, depth: Int): PerftResult = {
    val result: PerftResult = new PerftResult
    if (depth == 0) {
      result.moveCount += 1
      return result
    }
    val moves: ListBuffer[GCoups] = gp.getCoupsValides()
    val it: Iterator[GCoups] = moves.iterator
    while (it.hasNext) {
      {
        val ui: UndoGCoups = new UndoGCoups
        if (gp.exec(it.next, ui)) {
          val subPerft = perft(gp, depth - 1)
          gp.unexec(ui)
          result.moveCount += subPerft.moveCount
        }
      }
    }
    result
  }
}

File: src/main/java/perft/PerftSpeedTest.java
package perft;

import perft.PerftCompare.PerftResult;
import position.*;

import java.io.IOException;
import java.util.List;

public class PerftSpeedTest {

    public static void main(String[] args) throws IOException {
        perftTest();
    }

    public static void perftTest() {
        //voir http://chessprogramming.wikispaces.com/Perft+Results     
        String f = ICodage.FEN_INITIALE();
        GPositionS gp = FenToGPosition.toGPosition(f);
        int max_depth = 6;
        double t0 = System.nanoTime();
        for (int depth = 1; depth <= max_depth; depth++) {
            PerftResult res = perft(gp, depth);
            double t1 = System.nanoTime();
            System.out.println("Depth " + depth + " : " + (t1 - t0) / 1000000000 + " sec");
            System.out.println("Count = " + res.moveCount);
        }
//        res = perft(gp, 2);
//        res = perft(gp, 3);
//        res = perft(gp, 4);
//        res = perft(gp, 5);
//        res = perft(gp, 6);
    }

    public static PerftResult perft(GPositionS gp, int depth) {

        PerftResult result = new PerftResult();
        if (depth == 0) {
            result.moveCount++;
            return result;
        }
        List<GCoups> moves = gp.getCoupsValides();
        for (GCoups move : moves) {
            UndoGCoups ui = new UndoGCoups();
            if (gp.exec(move, ui)) {
                PerftResult subPerft = perft(gp, depth - 1);
                gp.unexec(ui);
                result.moveCount += subPerft.moveCount;
            }
        }
        return result;
    }
}
